### PR TITLE
Fast JIT call_indirect optimization

### PR DESCRIPTION
**Pull request by chenyuyang2022**

### CODE REVIEW

**Comment by wenyongh** [comment]
@chenyuyang2022 I added some comments in your last PR #1442

**Comment by chenyuyang2022** [comment]
Thank you Wenyong,  I‘ll revise it.

**Comment by chenyuyang2022** [comment]
> @chenyuyang2022 I added some comments in your last PR #1442

Thank you Wenyong, I'll revise it in this PR.

**Comment by chenyuyang2022** [comment]
https://github.com/bytecodealliance/wasm-micro-runtime/pull/1442#discussion_r960335912

@wenyongh I got an error after replacing this part with your code.

VOIDfast jit compilation failed: unknown error
WASM module load failed: fast jit compilation failed

File: core/iwasm/fast-jit/fe/jit_emit_function.c [hunk 1]
```diff
@@ -331,10 +331,53 @@ jit_compile_op_call_indirect(JitCompContext *cc, uint32 type_idx,
...
+    GEN_INSN(MUL, offset, elem_idx_long, NEW_CONST(I64, sizeof(uint32)));
+    
+    func_idx = jit_cc_new_reg_I32(cc);
+    tbl_data = get_table_data_reg(jit_frame, tbl_idx);
+    GEN_INSN(LDI32, func_idx, tbl_data, offset);
+    // offset = jit_cc_new_reg_I32(cc);
+    // func_idx = jit_cc_new_reg_I32(cc);
+    // tbl_data = get_table_data_reg(jit_frame, tbl_idx);
+
+    // GEN_INSN(SHL, offset, elem_idx, 2);
```
Comment (wenyongh) [inline core/iwasm/fast-jit/fe/jit_emit_function.c:364]: Try `GEN_INSN(SHL, offset, elem_idx, NEW_CONST(I32, 2))` and replace L351 to L359 with the lines commented?
If it is still error, could you please dump the JIT IR of the first pass?

Comment (lum1n0us) [reply core/iwasm/fast-jit/fe/jit_emit_function.c:364]: `SHL` requires a fixed `RCX` ...

Comment (wenyongh) [reply core/iwasm/fast-jit/fe/jit_emit_function.c:364]: x86 SHL instruction can be `SHL r/m32, imm8`, the second operand can be an immediate value:
https://www.felixcloutier.com/x86/sal:sar:shl:shr

We also generate the similar JIT IR in the frontend, and translate it into `SHL r/m32, imm8` instruction in the backend:
https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/core/iwasm/fast-jit/fe/jit_emit_numberic.c#L1000-L1003

https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/core/iwasm/fast-jit/cg/x86-64/jit_codegen_x86_64.cpp#L3837-L3843

Comment (chenyuyang2022) [reply core/iwasm/fast-jit/fe/jit_emit_function.c:364]:
```
offset = jit_cc_new_reg_I32(cc);
func_idx = jit_cc_new_reg_I32(cc);
tbl_data = get_table_data_reg(jit_frame, tbl_idx);
GEN_INSN(SHL, offset, elem_idx, NEW_CONST(I32, 2));
// GEN_INSN(SHL, offset, elem_idx, 2);
 GEN_INSN(LDI32, func_idx, tbl_data, offset);
```

@wenyongh Still have problems:

i3fast jit compilation failed: unknown error
WASM module load failed: fast jit compilation failed

---
JIT.COMPILER.DUMP: PASS_NO=2 PREV_PASS=update_cfg

; Function: main
; Constant table sizes: VOID=0 I32=1 I64=2 F32=0 F64=0 V64=0 V128=0 V256=0
; Label number: 8
; Instruction number: 67
; Register numbers: VOID=0 I32=17 I64=39 F32=16 F64=16 V64=0 V128=0 V256=0
; Label annotations: basic_block begin_bcip end_bcip
; Instruction annotations:
; Register annotations: def_insn

L0:
    ; PREDS()
    ; BEGIN_BCIP=0x0000 END_BCIP=0x0000
    LDPTR           I26, I15, 0xc90
    LDPTR           I27, I15, 0xc88
    ADD             I29, I26, 0x170L
    CMP             i6, I29, I27
    BGTU            i6, L6, VOID
    SUB             I28, I29, 0xb8L
    STPTR           I28, I15, 0xc90
    ADD             I30, I26, 0x58L
    STPTR           I30, I26, 0x30
    STPTR           I0, I26, 0x0
    STPTR           I26, I15, 0x68
    MOV             I0, I26
    JMP             L2
    ; SUCCS(L2)

L2:
    ; PREDS(L0)
    ; BEGIN_BCIP=0x0043 END_BCIP=0x004c
    LDI32           i10, I0, 0x50
    LDI32           i11, I0, 0x54
    LDPTR           I16, I15, 0x10
    LDPTR           I23, I16, 0x38
    LDPTR           I24, I23, 0x0
    LDI32           i9, I24, 0x4
    CMP             i6, 0x0, i9
    BGEU            i6, L4, VOID
    ADD             I25, I24, 0xcL
    SHL             i12, 0x0, 0x2
    LDI32           i13, I25, i12
    CMP             i6, i13, 0xffffffff
    BEQ             i6, L5, VOID
    LDI32           i14, I16, 0x10
    CMP             i6, i13, i14
    BGTU            i6, L3, VOID
    STI32           i11, I0, 0x10c
    STI32           i10, I0, 0x108
    ADD             I31, I0, 0x58L
    STPTR           I31, I0, 0x30
    ADD             I32, I0, 0x108L
    MOV             I5, I15
    I32TOI64        I6, 0x0
    I32TOI64        I4, 0x0
    I32TOI64        I3, 0x1
    I32TOI64        I8, 0x2
    MOV             I9, I32
    CALLNATIVE      i1, 0x55d86a934f8bL, I5, I6, I4, I3, I8, I9
    MOV             i15, i1
    AND             i15, i15, 0xff
    CMP             i6, i15, 0x0
    BEQ             i6, L7, VOID
    LDI32           i16, I32, 0x0
    LDPTR           I33, I0, 0x0
    LDPTR           I34, I33, 0x30
    ADD             I34, I34, 0x4L
    STPTR           I34, I33, 0x30
    STPTR           I0, I15, 0xc90
    STPTR           I33, I15, 0x68
    MOV             I0, I33
    RETURNBC        0x0, i16, VOID
    ; SUCCS()

L3:
    ; PREDS()
    ; BEGIN_BCIP=0x0000 END_BCIP=0x0000
    LDPTR           I35, I15, 0x10
    CALLNATIVE      VOID, 0x55d86a95f970L, I35, 0x7
    RETURN          0x1
    ; SUCCS()

L4:
    ; PREDS()
    ; BEGIN_BCIP=0x0000 END_BCIP=0x0000
    LDPTR           I36, I15, 0x10
    CALLNATIVE      VOID, 0x55d86a95f970L, I36, 0x8
    RETURN          0x1
    ; SUCCS()

L5:
    ; PREDS()
    ; BEGIN_BCIP=0x0000 END_BCIP=0x0000
    LDPTR           I37, I15, 0x10
    CALLNATIVE      VOID, 0x55d86a95f970L, I37, 0x9
    RETURN          0x1
    ; SUCCS()

L6:
    ; PREDS()
    ; BEGIN_BCIP=0x0000 END_BCIP=0x0000
    LDPTR           I38, I15, 0x10
    CALLNATIVE      VOID, 0x55d86a95f970L, I38, 0x10
    RETURN          0x1
    ; SUCCS()

L7:
    ; PREDS()
    ; BEGIN_BCIP=0x0000 END_BCIP=0x0000
    RETURN          0x1
    ; SUCCS()

L1:
    ; PREDS()
    ; BEGIN_BCIP=0x0000 END_BCIP=0x0000
    ; SUCCS()

Comment (wenyongh) [reply core/iwasm/fast-jit/fe/jit_emit_function.c:364]: Yes, it is caused by `LDI32 i13, I25, i12`, i12 should not be int32 register, we should converted it to int64 register before LDI32. Try:
```C
    if (UINTPTR_MAX == UINT64_MAX) {
        JitReg offset_i32 = jit_cc_new_reg_I32(cc);
        offset = jit_cc_new_reg_I64(cc);
        GEN_INSN(SHL, offset_i32, elem_idx, NEW_CONST(I32, 2));
        GEN_INSN(I32TOI64, offset, offset_i32);
    }
    else {
        offset = jit_cc_new_reg_I32(cc);
        GEN_INSN(SHL, offset, elem_idx, NEW_CONST(I32, 2));
    }
    func_idx = jit_cc_new_reg_I32(cc);
    tbl_data = get_table_data_reg(jit_frame, tbl_idx);
    GEN_INSN(LDI32, func_idx, tbl_data, offset);
```

Comment (chenyuyang2022) [reply core/iwasm/fast-jit/fe/jit_emit_function.c:364]: It works. Thanks.

Comment (wenyongh) [reply core/iwasm/fast-jit/fe/jit_emit_function.c:364]: NP. Could you change the destination branch to `dev/fast_jit`?

Comment (chenyuyang2022) [reply core/iwasm/fast-jit/fe/jit_emit_function.c:364]: Sure.

**Comment by chenyuyang2022** [comment]
@wenyongh @lum1n0us 
我目前增加了三个block_import, block_nonimport和func_return三个BB，后续的结果处理等部分还没有写。方便的时候可以帮我看一下当前的部分存在哪些问题吗？谢谢~

**Pull request closed without merging**